---
title: "PigLungCode"
output: html_document
---

```{r, setup and loading packages}
source("http://bioconductor.org/biocLite.R")
biocLite(suppressUpdates = FALSE)
biocLite("ShortRead", suppressUpdates = FALSE)

biocLite("devtools")
#install.packages("Rcpp")
library("devtools")
#devtools::install_github("benjjneb/dada2")
library("dada2")
require("vegan")
#install.packages("dplyr")
library(dplyr)
library(tidyr)
library(phyloseq)
library(ggplot2)
#install.packages("picante")
library(picante)
library(stringr)
library(grid)
#install.packages("gridExtra")
library(gridExtra)
#install.packages("phangorn")
#install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
library(pairwiseAdonis)
install.packages("RColorBrewer")
library(RColorBrewer)

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#iocManager::install("DECIPHER")
#BiocManager::install("dada2", version = "3.10")

##One note, in the metadata table, the sample ID for WCF_T_1B_S93 seems to indicate it should be a tissue sample; however, the incorrect ID was entered when submitting for sequencing. A review of the lab notebook correctly identifys the sample as a Swab sample is correct in the metadata table. 
```

##Plate 1 and DADA2 Pipeline
```{r, set and check filepath, get list of sample names}
path = "~/PigLungs/MiseqFiles"
list.files(path)

fnFs = sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs = sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
sample.names = sapply(strsplit(basename(fnFs),"_L001"), `[`,1)
```

```{r, plot the quality profile for the forward reads}
plotQualityProfile(fnFs[1:20],aggregate = TRUE)
```

```{r, plot the quality profile for the reverse reads}
plotQualityProfile(fnRs[1:20], aggregate = TRUE)
```

```{r, set file path for where filtered reads will go}
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
```

```{r, filter and trim reads, using standard calls and trimming 20 off each end to remove primers}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,200),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, trimLeft = 20,
              compress=TRUE, multithread=FALSE)

head(out)
```

```{r, generate error function, dereplicating, and merging reads}
errF = learnErrors(filtFs, multithread = FALSE)
errR = learnErrors(filtRs, multithread = FALSE)
names(filtFs) = sample.names
names(filtRs) = sample.names
mergers <- vector("list", length(sample.names))
names(mergers) <- sample.names

for(sam in sample.names) {
  cat("Processing:", sam, "\n")
    derepF <- derepFastq(filtFs[[sam]])
    ddF <- dada(derepF, err=errF, multithread=FALSE)
    derepR <- derepFastq(filtRs[[sam]])
    ddR <- dada(derepR, err=errR, multithread=FALSE)
    merger <- mergePairs(ddF, derepF, ddR, derepR)
    mergers[[sam]] <- merger
}
```

```{r, making a sequence table}
seqtab = makeSequenceTable(mergers)
table(nchar(getSequences(seqtab)))
#Too many reads of lengths outside of anticipated amplicon size so will be removing all reads that don't fall within 250 to 260 bp
ncol(seqtab)
ncol(seqtab2)
seqtab2=seqtab[,nchar(colnames(seqtab)) %in% seq(250,260)]
require(dada2)
table(nchar(getSequences(cell.seqtabnochim)))

#Reduced number of SVs from 24109 to 10154

saveRDS(seqtab2, "~/PigLungs/R1seqtab2.RDS") #saving file to join with second run
saveRDS(seqtab.nochim, "~/PigLungs/R1seqtab.nochim.RDS")

#Following steps were performed to ensure ASVs were kept throughout single run and not completely lost
seqtab.nochim = removeBimeraDenovo(seqtab2, method="consensus", multithread = FALSE, verbose = TRUE)

getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(mergers, getN), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
colSums(track)

taxa = assignTaxonomy(seqtab.nochim, multithread = FALSE)
taxa = addSpecies(taxa, )

```

##Begin Plate 2 Analysis
```{r, for Plate 2}
path = "~/PigLungs/MiseqFiles2"
list.files(path)

fnFs = sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs = sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
sample.names = sapply(strsplit(basename(fnFs),"_L001"), `[`,1)
```

```{r, plot the quality profile for the forward reads}
plotQualityProfile(fnFs[1:20],aggregate = TRUE)
```

```{r, plot the quality profile for the reverse reads}
plotQualityProfile(fnRs[1:20], aggregate = TRUE)
```

```{r, set file path for where filtered reads will go}
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
```

```{r, filter and trim reads, using standard calls and trimming 20 off each end to remove primers}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,200),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, trimLeft = 20,
              compress=TRUE, multithread=FALSE)

head(out)
```

```{r, generate error function, dereplicating, and merging reads}
errF = learnErrors(filtFs, multithread = FALSE, randomize = TRUE)

errR = learnErrors(filtRs, multithread = FALSE, randomize = TRUE)
names(filtFs) = sample.names
names(filtRs) = sample.names
mergers <- vector("list", length(sample.names))
names(mergers) <- sample.names

for(sam in sample.names) {
  cat("Processing:", sam, "\n")
    derepF <- derepFastq(filtFs[[sam]])
    ddF <- dada(derepF, err=errF, multithread=FALSE)
    derepR <- derepFastq(filtRs[[sam]])
    ddR <- dada(derepR, err=errR, multithread=FALSE)
    merger <- mergePairs(ddF, derepF, ddR, derepR)
    mergers[[sam]] <- merger
}
R2seqtab = makeSequenceTable(mergers)
```

```{r, making a sequence table}
table(nchar(getSequences(R2seqtab)))

#Too many reads of lengths outside of anticipated amplicon size so will be removing all reads that don't fall within 250 to 260 bp

R2seqtab2=seqtab[,nchar(colnames(R2seqtab)) %in% seq(250,260)]

saveRDS(R2seqtab2, "~/PigLungs/R2seqtab2.RDS")
R1seqtab2=readRDS("~/PigLungs/R1seqtab2.RDS")
table(nchar(getSequences(seqtab2)))

seq.all=mergeSequenceTables(R2seqtab2,R1seqtab2)

#Reduced number of SVs from 31191 to 18825

#saveRDS(seqtab, "~/PigLungs/R1seqtab2.RDS") #saving file to join with second run
#saveRDS(seqtab.nochim, "~/PigLungs/R1seqtab.nochim.RDS")

seqtab.nochim = removeBimeraDenovo(seq.all, method="consensus", multithread = FALSE, verbose = TRUE)

saveRDS(seqtab.nochim, "~/PigLungs/seq.nochim.RDS")

#Final total of 21453 ASVs

getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(mergers, getN), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
colSums(track)

sample.names=row.names(seqtab.nochim)
samplenameDF = as.data.frame(sample.names)

require(dada2)

taxa = assignTaxonomy(seqtab.nochim, "~/PigLungs/silva_nr_v132_train_set.fa.gz", multithread = FALSE)
saveRDS(taxa, "~/PigLungs/taxa.RDS")
```

##Creation of tree

```{r, make list of sequences in fasta file for tree making}
library(dada2)
require(phangorn)
require(DECIPHER)

seqtab.nochim = readRDS("~/PigLungs/seq.nochim.RDS")
uniques=getUniques(seqtab.nochim)
sequences=names(uniques)
uniquesToFasta(uniques,fout = "~/pigs.fasta", ids=sequences)
```

```{bash, make a tree with SEPP using gg backbone tree}
wget  "https://raw.github.com/smirarab/sepp-refs/master/gg/sepp-package.tar.bz"

tar xvfj sepp-package.tar.bz

cd sepp-package/sepp
python setup.py config -c

./sepp-package/run-sepp.sh pigs.fasta pigs
```

```{r, view tree file with ggtree}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ggtree")

library(ape)
pig.tree = read.tree(file = "~/PigLungs/pigs_placement.tog.relabelled.tre") %>%
  keep.tip(., sequences)
```

##Creation of Phyloseq Object and usage of Decontam for contaminating sequence removal

```{r, create phyloseq object}
source('http://bioconductor.org/biocLite.R')
biocLite('phyloseq')
require(phyloseq)
PigLungMetaData = read.csv("~/PigLungs/PigLungMetaData.csv")
rownames(PigLungMetaData)=PigLungMetaData$SampleName
SampleData=data.frame(PigLungMetaData)

ps = phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(SampleData), 
               tax_table(taxa))

saveRDS(ps, "~/PigLungs/ps.RDS")

ps = readRDS("~/PigLungs/ps.RDS")

ps1 = merge_phyloseq(ps, pig.tree)

saveRDS(ps1, "~/PigLungs/pstree.RDS")
```

```{r, usage of decontam to remove contaminating sequences}
library(devtools)
devtools::install_github("benjjneb/decontam")
library(decontam)
library(ggplot2)

#split ps object into different sample types based on kit or handling BloodTissue, PowerSoil, PowerWater, Bacteremia

ps.BT = subset_samples(ps, sample_data(ps)$ExtractionKit == "BloodTissue")

contamdf.freq = isContaminant(ps.BT, method="frequency", conc = "Qbit2", threshold = 0.2)

hist(contamdf.freq$p)

table(contamdf.freq$contaminant)
#Identified 27 contaminant taxa in Blood and Tissue kit at 0.2

#remove contaminant taxa
ps.BT.noncontam = prune_taxa(!contamdf.freq$contaminant, ps.BT)

#now on to the next kit
ps.PS = subset_samples(ps, sample_data(ps)$ExtractionKit == "PowerSoil")

contamdf.freq = isContaminant(ps.PS, method="frequency", conc = "Qbit2", threshold = 0.2)

hist(contamdf.freq$p)

table(contamdf.freq$contaminant)
#Identified 220 contaminant taxa in PowerSoil kit at 0.2

#remove contaminant taxa
ps.PS.noncontam = prune_taxa(!contamdf.freq$contaminant, ps.PS)

ps.PW = subset_samples(ps, sample_data(ps)$ExtractionKit == "PowerWater")

contamdf.freq = isContaminant(ps.PW, method="frequency", conc = "Qbit2", threshold = 0.2)

hist(contamdf.freq$p)

table(contamdf.freq$contaminant)
#Identified 478 contaminant taxa in PowerWater kit at 0.2

#remove contaminant taxa
ps.PW.noncontam = prune_taxa(!contamdf.freq$contaminant, ps.PW)

ps.Ba = subset_samples(ps, sample_data(ps)$ExtractionKit == "Bacteremia")

contamdf.freq = isContaminant(ps.Ba, method="frequency", conc = "Qbit2", threshold = 0.2)

hist(contamdf.freq$p)

table(contamdf.freq$contaminant)
#Identified 121 contaminant taxa in Bacteremia kit at 0.2

#remove contaminant taxa
ps.Ba.noncontam = prune_taxa(!contamdf.freq$contaminant, ps.Ba)

rm(ps.Ba, ps.BT, ps.PW, ps.PS)
ps.noncontam=merge_phyloseq(ps.BT.noncontam,ps.Ba.noncontam,ps.PW.noncontam,ps.PS.noncontam)
rm(ps.BT.noncontam,ps.Ba.noncontam,ps.PW.noncontam,ps.PS.noncontam)
ps.1=prune_species(speciesSums(ps.noncontam) > 0, ps.noncontam)
rm(ps.noncontam)
```

```{r, subset samples with PBS added}
ps.PBS = subset_samples(ps, sample_data(ps)$SampleType == "Lavage")
ps.PBS.p =prune_species(speciesSums(ps.PBS)>0, ps.PBS)

contamdf.freq = isContaminant(ps.PBS.p, method="frequency", conc = "Qbit2")

table(contamdf.freq$contaminant)

head(which(contamdf.freq$contaminant))

plot_frequency(ps, taxa_names(ps)[c(55,2)], conc = "Qbit2") + xlab("DNA Concentration")
```

##Data Analysis

```{r, make PCoA for all samples with unifrac and PERMANOVA}
#library(dplyr)
#library(ggplot2)
require(tidyr)
#ps.tree=readRDS("~/PigLungs/pstree.RDS")

#transform sample counts to relative abundance
ps.r=transform_sample_counts(ps.tree,function(x) x/sum(x))

#remove blank and NTC samples
ps.rncp=subset_samples(ps.r, sample_data(ps.r)$Sample != "NTC" & sample_data(ps.r)$SampleName != "FOF_AIR_BLANK_S92" & sample_data(ps.r)$SampleName != "NCAT_AIR_BLANK_S73") %>%
  prune_species(speciesSums(.) > 0, .) %>%
  subset_taxa(., Kingdom == "Bacteria")
rm(ps.r)

#check names to make sure they're removed
sample_names(ps.rncp)

#save ps object just in case
saveRDS(ps.rncp, "~/PigLungs/ps_ord.RDS")
ps.rncp = readRDS("~/PigLungs/ps_ord.RDS")

p=ordinate(ps.rncp, method = "PCoA", distance = "unifrac") %>%
plot_ordination(ps.rncp, ., type="samples", color = "Sample", shape = "Farm")+stat_ellipse()+theme_classic() 

p

#look at PERMANOVA to see if statistically different
metadf = data.frame(sample_data(ps.rncp))
metadf=unite(metadf, FarmSample, c(Farm, Sample), remove = F)
unifrac.dist = UniFrac(ps.rncp)

permanova = adonis(unifrac.dist ~ FarmSample, data = metadf)

permanova

pairwise.adonis(unifrac.dist, metadf$FarmSample) #pairwise testing with adonis shows that composition of each group is significantly different from each other, p-value bonferonni corrected
```

```{r, create ordination for env and lungs samples and PERMANOVA}
ps.Env = subset_samples(ps.rncp, sample_data(ps.rncp)$Sample == "Environment" & sample_data(ps.rncp)$Date != "Blank") %>%
  prune_species(speciesSums(.) > 0, .)

ordinate(ps.Env, method = "PCoA", distance = "unifrac") %>%
  plot_ordination(ps.Env, ., type="samples", color = "SampleType", shape = "Farm") + stat_ellipse()

ps.Pig = subset_samples(ps.rncp, sample_data(ps.rncp)$Sample == "Pig")

ordinate(ps.Pig, method = "PCoA", distance = "unifrac") %>%
  plot_ordination(ps.Pig, ., type="samples", color = "SampleType", shape = "Farm")+stat_ellipse()

Envmeta = data.frame(sample_data(ps.Env))

unifrac.dist = UniFrac(ps.Env)

permanova = adonis(unifrac.dist ~ Farm*SampleType, data = Envmeta) #permanova shows no significant differences (p=0.064) in farm*SampleType but there are significant differences in farm and sample type

permanova

Envmeta=unite(Envmeta, FarmSample, c(Farm, SampleType), remove = F)

library(pairwiseAdonis)

pairwise.adonis(unifrac.dist, Envmeta$SampleType)

Pigmeta = data.frame(sample_data(ps.Pig)) %>%
  unite(., FarmSample, c(Farm, SampleType), remove = F)

unifrac.dist = UniFrac(ps.Pig)

permanova = adonis(unifrac.dist ~ Farm*SampleType, data = Pigmeta)

permanova

pairwise.adonis(unifrac.dist, Pigmeta$Farm)

pairwise.adonis(unifrac.dist, Pigmeta$FarmSample) #Tissue samples are significantly different (adjusted p=0.015), Swab as well (p=0.015), and Lavage (p=0.015)

#keep the environment clean
rm(ps.Pig, ps.Env)
```

```{r, anosim results for Pig and Env Data, redo ENV to remove blanks}
ps.Env = subset_samples(ps.rncp, sample_data(ps.rncp)$Sample == "Environment" & sample_data(ps.rncp)$Date != "Blank") %>%
  prune_species(speciesSums(.) > 0, .)

ps.Pig = subset_samples(ps.rncp, sample_data(ps.rncp)$Sample == "Pig")

Envmeta = data.frame(sample_data(ps.Env)) %>%
  unite(., "FarmSample", c("Farm", "SampleType"),remove = F)

unifrac.dist = UniFrac(ps.Env)

set.seed(100)
anosim(unifrac.dist, Envmeta$SampleType, permutations = 9999) # R = 0.08, significance = 0.066

anosim(unifrac.dist, Envmeta$Farm, permutations = 9999) #R = 0.44, significance = 1e-4

EnvNMDS=metaMDS(unifrac.dist, k=3)

stressplot(EnvNMDS)

plot(EnvNMDS)

Pigmeta = data.frame(sample_data(ps.Pig)) %>%
  unite(., "FarmSample", c("Farm", "SampleType"), remove = F)

unifrac.dist = UniFrac(ps.Pig)

anosim(unifrac.dist, Pigmeta$Farm, permutations = 9999) #R = 0.52, significance = 1 e-4

anosim(unifrac.dist, Pigmeta$SampleType, permutations = 9999) #R = 0.40, significance = 1e-4
```

```{r, look at alpha diversity values for the samples}
#get Faith for NCAT
ps.alpha=subset_samples(ps.rncp, sample_data(ps.rncp)$Sample == "Pig") %>%
  prune_species(speciesSums(.) > 0, .)

sample = vegan_otu(ps.alpha)
tree = phy_tree(ps.alpha)

pd.result = pd(sample, tree)

pd.result$Samples = rownames(pd.result)

PD = separate(pd.result, Samples, into = c("Farm", "SampleType", NA, NA), sep = "_") %>% 
  mutate(SampleType = str_replace(SampleType, "T", "Tissue")) %>% 
  mutate(SampleType = str_replace(SampleType, "L", "Lavage")) %>% 
  mutate(SampleType = str_replace(SampleType, "S", "Swab"))

rm(sample, tree, pd.result)

g1 = ggplot(PD, aes(SampleType, PD, fill = Farm)) + geom_boxplot() +
  theme_minimal()+scale_fill_manual(values = c("#BF812D","darkslategray4"))+
  xlab(element_blank())

Shannon.result = estimate_richness(ps.alpha, measures = "Shannon")

Shannon.result$Samples = rownames(Shannon.result)

Shannon = separate(Shannon.result, Samples, into = c("Farm", "SampleType", NA, NA), sep = "_") %>% 
  mutate(SampleType = str_replace(SampleType, "T", "Tissue")) %>% 
  mutate(SampleType = str_replace(SampleType, "L", "Lavage")) %>% 
  mutate(SampleType = str_replace(SampleType, "S", "Swab"))

rm(Shannon.result)

g2=ggplot(Shannon, aes(SampleType, Shannon, fill = Farm)) + 
  geom_boxplot() +expand_limits(y=0)  + 
  theme_minimal()+scale_fill_manual(values = c("#BF812D","darkslategray4"))+
  xlab(element_blank())

grid.arrange(g1, g2, nrow = 1)

g=arrangeGrob(g1,g2,nrow=1)

ggsave(file="~/PigLungs/PublicationFigures/LungalphaDiv.PDF", g, width = 12, height = 6, units = "in")

#comparison of alpha diversities at each farm for the pig samples
fit = aov(PD ~ Farm*SampleType, data =  PD)
plot(fit)

#data looks right skewed in plot, so using Shapiro-Wilk test to determine if normally distributed
shapiro.test(PD$PD) #p<0.05 so data is not normally distrubted and ANOVA would be incorrectly applied
PD=unite(PD, FarmSample, c(Farm, SampleType), remove = F) #want to do farm by sample so making new column for kruskal test

kruskal.test(PD ~ FarmSample, data = PD)
pairwise.wilcox.test(PD$PD, PD$FarmSample, p.adjust.method = "BH") #only lavage (p = 0.033) and swab (p = 3.9e-5) samples between farms are different

#don't hate but I'm adding significance bars later in adobe acrobat 

kruskal.test(PD ~ SampleType, data = PD)
pairwise.wilcox.test(PD$PD, PD$SampleType, p.adjust.method = "BH") #indicates Tissue samples are different from lavage and swab (p < 0.05) but lavage and swab are not different

fit = aov(Shannon ~ Farm*SampleType, data = Shannon)
plot(fit)
shapiro.test(Shannon$Shannon) # p=0.22 so data is normally distributed

summary(fit) #indicates there is no significant difference by Farm, SampleType, or Farm*SampleType
```

```{r, look at alpha diversity values for the environment samples across farms}
#library(phyloseq)
#library(ggplot2)
#library(dplyr) 
#library(picante)
#library(tidyr)
#library(stringr)
#library(gridExtra)


ps.alpha=subset_samples(ps.rncp, sample_data(ps.rncp)$Sample == "Environment" & sample_data(ps.rncp)$Date != "Blank") %>%
  prune_species(speciesSums(.) > 0, .)

sample = vegan_otu(ps.alpha)
tree = phy_tree(ps.alpha)

pd.result = pd(sample, tree)

pd.result$Samples = rownames(pd.result)

PD = tidyr::separate(pd.result, Samples, into = c("Farm", "SampleType", NA, NA), sep = "_") %>% 
  mutate(SampleType = str_replace(SampleType, "W", "Water")) %>% 
  mutate(SampleType = str_replace(SampleType, "FWater", "Water")) %>%
  mutate(SampleType = str_replace(SampleType, "MWater", "Water")) %>% 
  mutate(SampleType = str_replace(SampleType, "FF", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType, "MF", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType, "FOOD1", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType,"FOOD2", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType,"FOOD3", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType, "AIR", "Air")) %>% 
  mutate(SampleType = str_replace(SampleType,"FS", "Soil")) %>% 
  mutate(SampleType = str_replace(SampleType,"MS", "Soil"))%>% 
  mutate(SampleType = str_replace(SampleType, "SOIL1", "Soil"))%>% 
  mutate(SampleType = str_replace(SampleType, "SOIL2", "Soil"))

rm(sample, tree, pd.result)

g1=ggplot(PD, aes(SampleType, PD, fill = Farm)) + geom_boxplot() + expand_limits(y=0) +
  theme_minimal()+scale_fill_manual(values = c("#BF812D","darkslategray4"))+
  xlab(element_blank())

Shannon.result = estimate_richness(ps.alpha, measures = "Shannon")

Shannon.result$Samples = rownames(Shannon.result)

Shannon = separate(Shannon.result, Samples, into = c("Farm", "SampleType", NA, NA), sep = "_") %>% 
  mutate(SampleType = str_replace(SampleType, "W", "Water")) %>% 
  mutate(SampleType = str_replace(SampleType, "FWater", "Water")) %>%
  mutate(SampleType = str_replace(SampleType, "MWater", "Water")) %>% 
  mutate(SampleType = str_replace(SampleType, "FF", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType, "MF", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType, "FOOD1", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType,"FOOD2", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType,"FOOD3", "Food")) %>% 
  mutate(SampleType = str_replace(SampleType, "AIR", "Air")) %>% 
  mutate(SampleType = str_replace(SampleType,"FS", "Soil")) %>% 
  mutate(SampleType = str_replace(SampleType,"MS", "Soil"))%>% 
  mutate(SampleType = str_replace(SampleType, "SOIL1", "Soil"))%>% 
  mutate(SampleType = str_replace(SampleType, "SOIL2", "Soil"))

rm(Shannon.result)

g2=ggplot(Shannon, aes(SampleType, Shannon, fill = Farm)) + geom_boxplot() + expand_limits(y=0) +
  theme_minimal()+scale_fill_manual(values = c("#BF812D","darkslategray4"))+
  xlab(element_blank())

g=grid.arrange(g1, g2, nrow = 1)

ggsave(file="~/PigLungs/PublicationFigures/EnvalphaDiv.PDF", g, width = 12, height = 6, units = "in")

fit = aov(PD ~ Farm*SampleType, data =  PD)
plot(fit)

#data looks right skewed in plot, so using Shapiro-Wilk test to determine if normally distributed
shapiro.test(PD$PD) #p<0.05 so data is not normally distrubted and ANOVA would be incorrectly applied
PD=unite(PD, FarmSample, c(Farm, SampleType), remove = F) #want to do farm by sample so making new column for kruskal test

kruskal.test(PD ~ FarmSample, data = PD) #p=0.26 so none of the medians are different between the Farm*SampleType

kruskal.test(PD ~ SampleType, data = PD) #p=0.16 so none of the medians are different between SampleTypes

pairwise.wilcox.test(PD$PD, PD$Farm, p.adjust.method = "BH") #p=0.079 so shows there is no difference between farms 

fit = aov(Shannon ~ Farm*SampleType, data = Shannon)
plot(fit)
shapiro.test(Shannon$Shannon) # p=0.1 so data is normally distributed

summary(fit) #indicates there is a significant differnce by SampleType with differences being Food-Air (0.03), Soil-Food (0.005), Water-Food (0.014)
TukeyHSD(fit)
```

```{r, creation of NCAT pig relative abundance chart, Plots were not used for publication, only posters}
subset_samples(ps.rncp, sample_data(ps.r)$Sample=="Pig" & sample_data(ps.r)$Farm == "NCAT") %>%
  prune_species(speciesSums(.) >0, .) %>%
  tax_glom(., taxrank = "Phylum") %>%
  prune_species(speciesSums(.) >2e-1, .) %>%
  plot_bar(., fill = "Phylum")

subset_samples(ps.rncp, sample_data(ps.r)$Sample=="Pig" & sample_data(ps.r)$Farm == "WCF") %>%
  prune_species(speciesSums(.) >0, .) %>%
  tax_glom(., taxrank = "Phylum") %>%
  prune_species(speciesSums(.) >2e-1, .) %>%
  plot_bar(., fill = "Phylum")
```

```{r, Lefse object for Pig swab and lavage samples}
ps.pig=subset_samples(ps.rncp, sample_data(ps.rncp)$Sample=="Pig") %>%
  prune_species(speciesSums(.) >0, .)

sample_names(ps.pig) #make sure we have the samples we want

lefse.pig = as.data.frame(cbind(t(otu_table(ps.pig, taxa_are_rows = FALSE)), tax_table(ps.pig)))

lefse.pig$Info=paste(lefse.pig$Kingdom, lefse.pig$Phylum, lefse.pig$Class, lefse.pig$Order, lefse.pig$Family,lefse.pig$Genus, sep = "|")
lefse.pig$Kingdom = NULL
lefse.pig$Phylum=NULL
lefse.pig$Class=NULL
lefse.pig$Order=NULL
lefse.pig$Family=NULL
lefse.pig$Genus=NULL

LEFseMeta = as.data.frame(sample_data(ps.pig)) %>%
  select(., c(SampleType,Farm,SampleName)) %>%
  mutate(SampleType = str_replace(SampleType, "Tissue", "Lower")) %>%
  mutate(SampleType = str_replace(SampleType, "Swab", "Upper")) %>%
  mutate(SampleType = str_replace(SampleType, "Lavage", "Upper")) %>%
  t(.) %>%
  as.data.frame(.) %>%
  mutate(., Info=row.names(.))

lefse.out=merge(lefse.pig, LEFseMeta, all=TRUE)
rm(lefse.pig, ps.pig)

write.csv(lefse.out, "~/PigLungs/lefsepig.csv", quote = FALSE)

#once I ran Lefse through galaxy, I removed all of the genera that weren't significant and imported the resulting excel tab

UpperLower = read.csv("~/PigLungs/LEfsEResultsUpperLower.csv") %>%
  mutate(., Genus = ï..Genus)

#set levels by LDA score
UpperLower$Genus=factor(UpperLower$Genus,levels=c("Delftia", "Yersinia", "Megasphaera", "Aerococcus", "Dictyoglomus", "Tepidiphilus", "Caldicellulosiruptor", "Escherichia_Shigella", "Streptococcus", "Anoxybacillus", "Psychrobacter", "Lachnospiraceae_XPB1014_group", "Lactobacillus", "Atopostipes", "Lachnospiraceae_AC2044_group", "Collinsella", "Fervidobacterium", "IheB3_7", "Bergeyella", "Peptoclostridium", "Oscillospira", "Ignavigranum", "dgA_11_gut_group"))

p=ggplot(UpperLower, aes(x=Genus,y=LDAScore))+geom_col(aes(fill=Location))+coord_flip()+theme_classic()

ggsave(file="~/PigLungs/PublicationFigures/lefse.PDF", p, width = 20, height = 20, units = "cm")
```

```{r, function for getting phyloseq object into usable format for SourceTracker}
#I found this online and for the life of me can't remember where it came from

function(ps, filePATH, filePREFIX, writeFASTA=TRUE, rename=FALSE, useREFSEQ=FALSE){
  
  #pull seqs from refseq slot or extract from ASV ID for fasta format
  if (isTRUE(useREFSEQ)){
    #from phyloseq refseq slot
    f.onames <- phyloseq::refseq(ps)
  } else {
    f.onames <- phyloseq::taxa_names(ps)
  }
  
  if (isTRUE(rename)){
    phyloseq::taxa_names(ps) <- paste("ASV", 1:length(phyloseq::taxa_names(ps)), sep = "")
    names(f.onames) <- paste0(">", phyloseq::taxa_names(ps))
  } else {
    names(f.onames) <- paste0(">", phyloseq::taxa_names(ps))
  }
  
  
  #generate asv table formatted for biom generation
  asv.tab <- format.ASV.tab(ps)
  suppressWarnings(asv.tab <- as.matrix(asv.tab))
  cb <- as.matrix(cbind(rownames(asv.tab), asv.tab))
  rcb <- as.matrix(rbind(colnames(cb), cb))
  rcb[1,1] <- "#ASVID"
  rownames(rcb) <- NULL
  colnames(rcb) <- NULL
  
  #generate tax table formatted for biom generation
  tax.tab <- as.data.frame(phyloseq::tax_table(ps))
  tax.tab$taxonomy <- tidyr::unite_(tax.tab, "out", c(colnames(tax.tab)), sep = ";")
  cbt <- as.matrix(cbind(rownames(tax.tab), tax.tab$taxonomy))
  rcbt <- as.matrix(rbind(c("#ASVID", "taxonomy"), cbt))
  rownames(cbt) <- NULL
  colnames(cbt) <- NULL
  
  #generate sampledf table formatted for biom generation
  samdf <- suppressWarnings(as.matrix(phyloseq::sample_data(ps)))
  cbs <- as.matrix(cbind(rownames(samdf), samdf))
  rcbs <- as.matrix(rbind(colnames(cbs), cbs))
  rcbs[1,1] <- "#SampleID"
  rownames(rcbs) <- NULL
  colnames(rcbs) <- NULL
  
  #create output string
  if (isTRUE(writeFASTA)){
    fa <- print(paste0(filePATH, filePREFIX, "_ASVs.fasta"))
  }
  otb <- print(paste0(filePATH, filePREFIX, "_ASV_table.txt"))
  ttb <- print(paste0(filePATH, filePREFIX, "_ASV_taxonomy.txt"))
  stb <- print(paste0(filePATH, filePREFIX, "_sample_data.txt"))
  
  
  #write output
  #ASV fasta 
  if (isTRUE(writeFASTA)){
    write.table(x = f.onames, file = fa, quote = FALSE, sep = "\n", col.names = FALSE)
  }
  #asv.tab
  write.table(x = rcb, file = otb, row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t")
  #tax.tab
  write.table(x = rcbt, file = ttb, row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t")
  #sampledf
  write.table(x = rcbs, file = stb, row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t")
  
  #return phyloseq object with taxa renamed to ASV1, etc., if desired
  if (isTRUE(rename)){
    return(ps)
  }
}

```

```{r,SourceTracker for NCAT}
#remove all samples that aren't NCAT or what we care about
ps.NCAT = subset_samples(ps.prok, sample_data(ps.prok)$Farm == "NCAT" & sample_data(ps.prok)$SampleType != "Tissue" & sample_data(ps.prok)$SampleType != "Swab" & sample_data(ps.prok)$SampleSubtype != "NTC" & sample_data(ps.prok)$SampleSubtype != "Blank")

write.dataset(ps.NCAT, "~/PigLungs/sourcetracker", "NCAT_lavage_sourcetracker")

ps.NCAT_Swab = subset_samples(ps.prok, sample_data(ps.prok)$Farm == "NCAT" & sample_data(ps.prok)$SampleType != "Tissue" & sample_data(ps.prok)$SampleType != "Lavage" & sample_data(ps.prok)$SampleSubtype != "NTC" & sample_data(ps.prok)$SampleSubtype != "Blank")

write.dataset(ps.NCAT_Swab, "~/PigLungs/sourcetracker", "NCAT_swab_sourcetracker")

ps.NCAT_Tissue = subset_samples(ps.prok, sample_data(ps.prok)$Farm == "NCAT" & sample_data(ps.prok)$SampleType != "Swab" & sample_data(ps.prok)$SampleType != "Lavage" & sample_data(ps.prok)$SampleSubtype != "NTC" & sample_data(ps.prok)$SampleSubtype != "Blank")

write.dataset(ps.NCAT_Tissue, "~/PigLungs/sourcetracker", "NCAT_tissue_sourcetracker")

ps.NCAT = subset_samples(ps.prok, sample_data(ps.prok)$Farm == "NCAT")  

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "1") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s1")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "2") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s2")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "3") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s3")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "4") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s4")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "5") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s5")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "6") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s6")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "7") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s7")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "8") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s8")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "9") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s9")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "10") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s10")

##No lavage sample for 11 so excluded from analysis

#ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "11") %>%
#prune_species(speciesSums(.) > 0, .)

#write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s11")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "12") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s12")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "13") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s13")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "14") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s14")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "15") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s15")

ps.NCAT_s1=subset_samples(ps.NCAT,sample_data(ps.NCAT)$Sample == "Environment" | sample_data(ps.NCAT)$PigID == "16") %>%
prune_species(speciesSums(.) > 0, .)

write.dataset(ps.NCAT_s1, "~/PigLungs/sourcetracker/", "NCAT_s16")
```

```{bash, bash input for sourcetracker for NCAT}
#Need to download sourcetracker_for_qiime.r
#set filepaths in environment
echo "export SOURCETRACKER_PATH=$HOME/Documents/PigLungs/sourcetracker" >> ~/.bash_profile
source ~/.bash_profile

#Didn't realize write.dataset would have some of the wrong column names and be missing a row after all the dataset were written so I did the following in excel
#need to set up sourcetracker software add source/sink row to sample_data and one extra row beginning with a # on ASV table
#environmental samples were source and pig samples were sinks

Rscript sourcetracker_for_qiime.r -i NCAT_lavage_sourcetracker_ASV_table.txt -m NCAT_lavage_sourcetracker_sample_data.txt -o lavage_full_NCAT -v -n 25

Rscript sourcetracker_for_qiime.r -i NCAT_tissue_sourcetracker_ASV_table.txt -m NCAT_tissue_sourcetracker_sample_data.txt -o tissue_full_NCAT -v -n 25

Rscript sourcetracker_for_qiime.r -i NCAT_swab_sourcetracker_ASV_table.txt -m NCAT_swab_sourcetracker_sample_data.txt -o swab_full_NCAT -v -n 25
```

```{r, aggregation of sourcetracker txt files into dataframe for NCAT to show average}
tissue_ST=read.table("~/PigLungs/sourcetracker/tissue_full_NCAT/sink_predictions.txt",header = T, sep = "\t") %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID)))

tissue_ST_all = tissue_ST %>%
  mutate(., sampletype = "tissue") %>%
  unite(., Group, group, sampletype) %>%
  select(., -SampleID)
  
tissue_ST_avg = tissue_ST %>%
  reshape2::melt(.) %>%
  group_by(group, variable) %>%
  filter(!(abs(value - median(value)) > 2*sd(value))) %>%
  summarise_each(funs(mean), value) %>%
  mutate(., sampletype = "tissue") %>%
  unite(., Group, sampletype, group, sep = "_") %>%
  group_by(.,Group) %>%
  mutate(., value = value*(1/sum(value)))%>%
  as.data.frame(.)

sum(tissue_ST_avg$value)
  
lavage_ST=read.table("~/PigLungs/sourcetracker/lavage_full_NCAT/sink_predictions.txt",header = T, sep = "\t") %>%
  mutate(.,Group = "lavage")

lavage_ST_all = lavage_ST %>%
  select(., -SampleID)

lavage_ST_avg=lavage_ST %>%
  reshape2::melt(.) %>%
  group_by(variable) %>%
  filter(!(abs(value - median(value)) > 2*sd(value))) %>%
  summarise_each(funs(mean), value)%>%
  mutate(., Group = "lavage") %>%
#since outliers were removed data no longer adds up to 1, so need to normalize each value to get to 1
  mutate(., value = value*(1/sum(value)))%>%
  as.data.frame(.)

#check that normalization worked
sum(lavage_ST_avg$value)

swab_ST=read.table("~/PigLungs/sourcetracker/swab_full_NCAT/sink_predictions.txt",header = T, sep = "\t") %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

swab_ST_all = swab_ST %>%
  mutate(., sampletype = "swab") %>%
  unite(., Group, group, sampletype) %>%
  select(., -SampleID)

swab_ST_avg = swab_ST %>%
  reshape2::melt(.) %>%
  group_by(group, variable) %>%
  filter(!(abs(value - median(value)) > 2*sd(value))) %>%
  summarise_each(funs(mean), value) %>%
  mutate(., sampletype = "swab") %>%
  unite(., Group, sampletype, group, sep = "_") %>%
  group_by(.,Group) %>%
  mutate(., value = value*(1/sum(value))) %>%
  as.data.frame(.)

sum(swab_ST_avg$value)

STDF_avg = rbind(swab_ST_avg, lavage_ST_avg, tissue_ST_avg) 

STDF_avg$variable=factor(STDF_avg$variable,levels=c("Food","Water", "Soil", "Unknown", "Air"))

STDF_avg$Group=factor(STDF_avg$Group,levels=c("lavage","swab_A", "swab_B", "tissue_B", "tissue_A"))

g=ggplot() + geom_bar(aes(y = value, x = Group, fill = variable), data = STDF_avg, stat ="identity")+scale_fill_manual(values=c( "green4", "royalblue", "brown", "grey52", "deepskyblue3"))+theme_bw()+ylab("Relative Abundance")+xlab("Sample Type")+ggtitle("NCAT")

ggsave(file="~/PigLungs/PublicationFigures/STAvgNCAT.PDF", g, width = 10, height = 10, units = "cm")

STDF_NCAT = bind_rows(swab_ST_all, lavage_ST_all, tissue_ST_all) %>%
  reshape2::melt(.)

STDF_NCAT$variable=factor(STDF_NCAT$variable,levels=c("Food","Water", "Soil", "Unknown", "Air"))

STDF_NCAT$Group=factor(STDF_NCAT$Group,levels=c("lavage","A_swab", "B_swab", "B_tissue", "A_tissue"))

g=ggplot(STDF_NCAT, aes(Group, value)) + facet_grid(. ~ variable) + geom_boxplot(aes(fill = variable)) + theme_bw()+scale_fill_manual(values=c( "green4", "royalblue", "brown", "grey52", "deepskyblue3"))+ylab("Relative Abundance")+xlab("Sample Type")+ggtitle("NCAT")

ggsave(file="~/PigLungs/PublicationFigures/STNCAT.PDF", g, width = 20, height = 10, units = "cm")
```

```{r, make ST files for FOF}
ps.FOF = subset_samples(ps.prok, sample_data(ps.prok)$Farm == "FOF" & sample_data(ps.prok)$SampleType != "Tissue" & sample_data(ps.prok)$SampleType != "Swab" & sample_data(ps.prok)$SampleName != "FOF_AIR_BLANK_S92")

write.dataset(ps.FOF, "~/PigLungs/sourcetracker/", "FOF_lavage_sourcetracker")

ps.FOF_Swab = subset_samples(ps.prok, sample_data(ps.prok)$Farm == "FOF" & sample_data(ps.prok)$SampleType != "Tissue" & sample_data(ps.prok)$SampleType != "Lavage" & sample_data(ps.prok)$SampleName != "FOF_AIR_BLANK_S92")

write.dataset(ps.FOF_Swab, "~/PigLungs/sourcetracker/", "FOF_swab_sourcetracker")

ps.FOF_Tissue = subset_samples(ps.prok, sample_data(ps.prok)$Farm == "FOF" & sample_data(ps.prok)$SampleType != "Swab" & sample_data(ps.prok)$SampleType != "Lavage" & sample_data(ps.prok)$SampleName != "FOF_AIR_BLANK_S92")

write.dataset(ps.FOF_Tissue, "~/PigLungs/sourcetracker/", "FOF_tissue_sourcetracker")
```

```{bash, bash input for sourcetracker for FOF}
#Need to download sourcetracker_for_qiime.r
#set filepaths in environment
echo "export SOURCETRACKER_PATH=$HOME/Documents/PigLungs/sourcetracker" >> ~/.bash_profile
source ~/.bash_profile

#Didn't realize write.dataset would have some of the wrong column names and be missing a row after all the dataset were written so I did the following in excel
#need to set up sourcetracker software add source/sink row to sample_data and one extra row beginning with a # on ASV table
#environmental samples were source and pig samples were sinks

Rscript sourcetracker_for_qiime.r -i FOF_lavage_sourcetracker_ASV_table.txt -m FOF_lavage_sourcetracker_sample_data.txt -o lavage_full_FOF -v -n 25

Rscript sourcetracker_for_qiime.r -i FOF_tissue_sourcetracker_ASV_table.txt -m FOF_tissue_sourcetracker_sample_data.txt -o tissue_full_FOF -v -n 25

Rscript sourcetracker_for_qiime.r -i FOF_swab_sourcetracker_ASV_table.txt -m FOF_swab_sourcetracker_sample_data.txt -o swab_full_FOF -v -n 25
```

```{r, aggregation of sourcetracker txt files into dataframe for FOF}
tissue_ST=read.table("~/PigLungs/sourcetracker/tissue_full_WCF/sink_predictions.txt",header = T, sep = "\t") %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID)))

tissue_ST_all = tissue_ST %>%
  mutate(., sampletype = "tissue") %>%
  unite(., Group, group, sampletype) %>%
  select(., -SampleID)
  
tissue_ST_avg = tissue_ST %>%
  reshape2::melt(.) %>%
  group_by(group, variable) %>%
  filter(!(abs(value - median(value)) > 2*sd(value))) %>%
  summarise_each(funs(mean), value) %>%
  mutate(., sampletype = "tissue") %>%
  unite(., Group, sampletype, group, sep = "_") %>%
  group_by(.,Group) %>%
  mutate(., value = value*(1/sum(value)))%>%
  as.data.frame(.)
  
lavage_ST=read.table("~/PigLungs/sourcetracker/lavage_full_WCF/sink_predictions.txt",header = T, sep = "\t") %>%
  mutate(.,Group = "lavage")

lavage_ST_all = lavage_ST %>%
  select(., -SampleID)

lavage_ST_avg=lavage_ST %>%
  reshape2::melt(.) %>%
  group_by(variable) %>%
  filter(!(abs(value - median(value)) > 2*sd(value))) %>%
  summarise_each(funs(mean), value)%>%
  mutate(., Group = "lavage") %>%
#since outliers were removed data no longer adds up to 1, so need to normalize each value to get to 1
  mutate(., value = value*(1/sum(value)))%>%
  as.data.frame(.)

swab_ST=read.table("~/PigLungs/sourcetracker/swab_full_WCF/sink_predictions.txt",header = T, sep = "\t") %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

swab_ST_all = swab_ST %>%
  mutate(., sampletype = "swab") %>%
  unite(., Group, group, sampletype) %>%
  select(., -SampleID)

swab_ST_avg = swab_ST %>%
  reshape2::melt(.) %>%
  group_by(group, variable) %>%
  filter(!(abs(value - median(value)) > 2*sd(value))) %>%
  summarise_each(funs(mean), value) %>%
  mutate(., sampletype = "swab") %>%
  unite(., Group, sampletype, group, sep = "_") %>%
  group_by(.,Group) %>%
  mutate(., value = value*(1/sum(value))) %>%
  as.data.frame(.)

STDF_avg = rbind(swab_ST_avg, lavage_ST_avg, tissue_ST_avg) 

STDF_avg$variable=factor(STDF_avg$variable,levels=c("Food","Water", "Soil", "Unknown", "Air"))

STDF_avg$Group=factor(STDF_avg$Group,levels=c("lavage","swab_A", "swab_B", "tissue_B", "tissue_A"))

g=ggplot() + geom_bar(aes(y = value, x = Group, fill = variable), data = STDF_avg, stat ="identity")+scale_fill_manual(values=c( "green4", "royalblue", "brown", "grey52", "deepskyblue3"))+theme_bw()+ylab("Relative Abundance")+xlab("Sample Type")+ggtitle("FOF")

ggsave(file="~/PigLungs/PublicationFigures/STAvgFOF.PDF", g, width = 10, height = 10, units = "cm")

STDF_FOF = bind_rows(swab_ST_all, lavage_ST_all, tissue_ST_all) %>%
  reshape2::melt(.)

STDF_FOF$variable=factor(STDF_FOF$variable,levels=c("Food","Water", "Soil", "Unknown", "Air"))

STDF_FOF$Group=factor(STDF_FOF$Group,levels=c("lavage","A_swab", "B_swab", "B_tissue", "A_tissue"))

g=ggplot(STDF_FOF, aes(Group, value)) + facet_grid(. ~ variable) + geom_boxplot(aes(fill = variable)) + theme_bw()+scale_fill_manual(values=c( "green4", "royalblue", "brown", "grey52", "deepskyblue3"))+ylab("Relative Abundance")+xlab("Sample Type")+ggtitle("FOF")

ggsave(file="~/PigLungs/PublicationFigures/STFOF.PDF", g, width = 20, height = 10, units = "cm")
```

```{r, aggregation of sourcetracker txt files into dataframe, swab/lavage analysis, I really should have written a function here and used purrr}
sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample1_NCAT/sink_predictions.txt", header = T, sep = "\t")

S1 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S1$group = paste0("tissue_", S1$group)
S1$SampleID = paste0("tissue_", S1$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample2_NCAT/sink_predictions.txt", header = T, sep = "\t")

S2 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S2$group = paste0("tissue_", S2$group)
S2$SampleID = paste0("tissue_", S2$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample3_NCAT/sink_predictions.txt", header = T, sep = "\t")

S3 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S3$group = paste0("tissue_", S3$group)
S3$SampleID = paste0("tissue_", S3$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample4_NCAT/sink_predictions.txt", header = T, sep = "\t")

S4 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S4$group = paste0("tissue_", S4$group)
S4$SampleID = paste0("tissue_", S4$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample5_NCAT/sink_predictions.txt", header = T, sep = "\t")

S5 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S5$group = paste0("tissue_", S5$group)
S5$SampleID = paste0("tissue_", S5$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample6_NCAT/sink_predictions.txt", header = T, sep = "\t")

S6 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID)))

S6$group = paste0("tissue_", S6$group)
S6$SampleID = paste0("tissue_", S6$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample7_NCAT/sink_predictions.txt", header = T, sep = "\t")

S7 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S7$group = paste0("tissue_", S7$group)
S7$SampleID = paste0("tissue_", S7$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample8_NCAT/sink_predictions.txt", header = T, sep = "\t")

S8 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S8$group = paste0("tissue_", S8$group)
S8$SampleID = paste0("tissue_", S8$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample9_NCAT/sink_predictions.txt", header = T, sep = "\t")

S9=sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S9$group = paste0("tissue_", S9$group)
S9$SampleID = paste0("tissue_", S9$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample10_NCAT/sink_predictions.txt", header = T, sep = "\t")

S10 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S10$group = paste0("tissue_", S10$group)
S10$SampleID = paste0("tissue_", S10$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample12_NCAT/sink_predictions.txt", header = T, sep = "\t")

S12 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S12$group = paste0("tissue_", S12$group)
S12$SampleID = paste0("tissue_", S12$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample13_NCAT/sink_predictions.txt", header = T, sep = "\t")

S13 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S13$group = paste0("tissue_", S13$group)
S13$SampleID = paste0("tissue_", S13$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample14_NCAT/sink_predictions.txt", header = T, sep = "\t")

S14 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S14$group = paste0("tissue_", S14$group)
S14$SampleID = paste0("tissue_", S14$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample15_NCAT/sink_predictions.txt", header = T, sep = "\t")

S15 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S15$group = paste0("tissue_", S15$group)
S15$SampleID = paste0("tissue_", S15$SampleID)

sample = read.table("~/PigLungs/sourcetracker/tissue_ST_from_swablavage/lavage_Sample16_NCAT/sink_predictions.txt", header = T, sep = "\t")

S16 = sample %>% 
  separate(SampleID, c("farm", "sampletype", "SampleID", "plateID"), remove=T) %>%
  select(., -c(farm, sampletype, plateID)) %>%
  mutate(group = substr(SampleID, nchar(SampleID)-0, nchar(SampleID))) 

S16$group = paste0("tissue_", S16$group)
S16$SampleID = paste0("tissue_", S16$SampleID)

SLDF = bind_rows(S1, S2, S3, S4, S5, S6, S7, S8, S9, S10, S12, S13, S14, S15, S16) %>%
  reshape2::melt(.)%>%
  group_by(group, variable) %>%
  filter(!(abs(value - median(value)) > 2*sd(value))) %>%
  summarise_each(funs(mean), value) %>%
  group_by(group) %>%
  mutate(., value = value*(1/sum(value)))%>%
  as.data.frame(.)
  
sum(SLDF$value)

SLDF$variable=factor(SLDF$variable,levels=c("Food","Water", "Soil", "Unknown", "Air", "Swab.Lavage"))

g=ggplot() + geom_bar(aes(y = value, x = group, fill = variable), data = SLDF, stat ="identity")+scale_fill_manual(values=c( "green4", "royalblue", "brown", "grey52", "deepskyblue3", "chocolate"))+ theme_bw()+ylab("Relative Abundance")+xlab("Sample Type")+ggtitle("NCAT")

ggsave(file="~/PigLungs/PublicationFigures/STforSLAvg.PDF", g, width = 5, height = 10, units = "cm")

SLDF2 = bind_rows(S1, S2, S3, S4, S5, S6, S7, S8, S9, S10, S12, S13, S14, S15, S16) %>%
  reshape2::melt(.)

SLDF2$variable=factor(SLDF2$variable,levels=c("Food","Water", "Soil", "Unknown", "Air", "Swab.Lavage"))

g=ggplot(SLDF2, aes(variable, value)) + facet_grid(. ~ variable) + geom_boxplot(aes(fill = variable)) +facet_grid(~group)+ theme_bw()+scale_fill_manual(values=c( "green4", "royalblue", "brown", "grey52", "deepskyblue3", "chocolate"))+ylab("Relative Abundance")+xlab("Sample Type")+ggtitle("NCAT")

ggsave(file="~/PigLungs/PublicationFigures/STforSL.PDF", g, width = 20, height = 10, units = "cm")
```


